### PR TITLE
Log Refactor [10222]

### PR DESCRIPTION
**Pull request by jparisu**
Set `LOG_NO_INFO` default ON.
`LOG_NO_INFO` = OFF compile LogInfo and queue the info (show info depends on verbosity and filter).

Signed-off-by: jparisu <javierparis@eprosima.com>

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3496/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/685/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3108/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3527/)

**Comment by jparisu** [comment]
Tests `LogMacrosInternalDebugOffTests` are disabled (commented in *CMakeLists.txt*) because it is a special case. This test will fail in compilation if test fails.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3508/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/697/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3119/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3539/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3513/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/703/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3124/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3545/)

**Review by rsanchez15 — CHANGES_REQUESTED** [review]
Partial review pending verification of all compilation cases.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3524/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/714/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3132/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3556/)

File: CMakeLists.txt [hunk 1]
```diff
@@ -197,7 +197,7 @@ endif()
 ###############################################################################
 
 if(EPROSIMA_BUILD AND NOT EPROSIMA_INSTALLER_MINION)
-    set(INTERNAL_DEBUG ON)
+    option(INTERNAL_DEBUG ON)
```
[the remark this answers is not in the document]
Comment (jparisu) [reply CMakeLists.txt:200]: It is not a duplicity. The second one set it to OFF in case it has not been set yet.
However, I also think it would be better to unify both in the same CMakeLists.

File: CMakeLists.txt [hunk 2]
```diff
@@ -278,6 +278,37 @@ if(LOG_CONSUMER_DEFAULT STREQUAL "STDOUTERR")
     set(STDOUTERR_LOG_CONSUMER ON)
 endif()
 
+###############################################################################
+# Log Info default setup
+###############################################################################
+if(CMAKE_BUILD_TYPE STREQUAL "Debug")
+    option(LOG_NO_INFO "Do not compile Info Log level" OFF)
```
[the remark this answers is not in the document]
Comment (jparisu) [reply CMakeLists.txt:285]: I do not see difference in the behaviour.
I think it is better this way as the defaul value is different depending the *build type*.
The other way around would suggest that it has a default value that is changed in a specific *build type* (almost the same, but I still think the first interpretation describes better the behaviour).

File: test/unittest/logging/log_macros/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,154 @@
+# Copyright 2016 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
[the remark this answers is not in the document]
Comment (jparisu) [reply test/unittest/logging/log_macros/CMakeLists.txt:1]: :+1:

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply CMakeLists.txt:285]: How about [CMakeDependentOption](https://cmake.org/cmake/help/latest/module/CMakeDependentOption.html)

File: test/unittest/logging/log_macros/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            )
+
+        # MACROS LOG TEST ALL
+        set(LOGMACROS_ALL_TEST_SOURCE LogMacrosAllActiveTests.cpp)
+
+        set(LOGMACROS_SOURCE
+            ${LOG_COMMON_SOURCE}
+            ${LOGMACROS_ALL_TEST_SOURCE})
+
+        include_directories(mock/)
```
[the remark this answers is not in the document]
Comment (jparisu) [reply test/unittest/logging/log_macros/CMakeLists.txt:38]: True, this should not be correct but it was working fine. I change it along the file.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply CMakeLists.txt:285]: Forget it, that would not make the option available when the condition is false

File: test/unittest/logging/log_macros/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+            ${PROJECT_SOURCE_DIR}/include ${PROJECT_BINARY_DIR}/include)
+        target_link_libraries(LogMacrosDefaultTests ${GTEST_LIBRARIES} ${MOCKS}
+            $<$<BOOL:${WIN32}>:iphlpapi$<SEMICOLON>Shlwapi>
+            )
+        add_gtest(LogMacrosDefaultTests SOURCES ${LOGMACROS_DEFAULT_TEST_SOURCE})
+
+
+        # MACROS LOG TEST INTERNALDEBUG
+        # This test is commented as it has a special way of failure
+        # In case this test fails, it fails de compilation
```
[the remark this answers is not in the document]
Comment (jparisu) [reply test/unittest/logging/log_macros/CMakeLists.txt:145]: @MiguelCompany approved the implementation of this test.

File: test/unittest/logging/log_macros/LogMacrosAllActiveTests.cpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#ifdef HAVE_LOG_NO_INFO
+#undef HAVE_LOG_NO_INFO
+#endif // HAVE_LOG_NO_INFO
+#define HAVE_LOG_NO_INFO 0
```
[the remark this answers is not in the document]
Comment (jparisu) [reply test/unittest/logging/log_macros/LogMacrosAllActiveTests.cpp:20]: It is a negative variable. 0 means NO `LOG_NO_*` so all log levels are active. Maybe the name of the test should change, but I do not know a better name. In the meanwhile I will add a comment to specify the behaviour.

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (rsanchez15) [reply CMakeLists.txt:285]: Yes, but you are giving different descriptions to the same CMake option. What I mean is that the option description must be unique since the option means only one thing, not two. It is the behavior what is different depending on the value of the option.

File: test/unittest/logging/log_macros/LogMacrosAllActiveTests.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    MockConsumer* mockConsumer;
+
+    const uint32_t AsyncTries = 5;
+    const uint32_t AsyncWaitMs = 25;
+
+    std::vector<Log::Entry> HELPER_WaitForEntries(
+            uint32_t amount);
+};
+
+std::vector<Log::Entry> LogMacrosAllActiveTests::HELPER_WaitForEntries(
```
[the remark this answers is not in the document]
Comment (jparisu) [reply test/unittest/logging/log_macros/LogMacrosAllActiveTests.cpp]: :+1:

File: test/unittest/logging/log_macros/LogMacrosAllActiveTests.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+#include <fastdds/dds/log/OStreamConsumer.hpp>
+#include <fastdds/dds/log/StdoutConsumer.hpp>
+#include <fastdds/dds/log/StdoutErrConsumer.hpp>
+#include "../mock/MockConsumer.h"
+#include <gtest/gtest.h>
+
+using namespace eprosima::fastdds::dds;
+using namespace std;
+
+class LogMacrosAllActiveTests : public ::testing::Test
```
[the remark this answers is not in the document]
Comment (jparisu) [reply test/unittest/logging/log_macros/LogMacrosAllActiveTests.cpp]: :+1:

File: CMakeLists.txt [hunk 9]
```diff
@@ -278,6 +277,37 @@ if(LOG_CONSUMER_DEFAULT STREQUAL "STDOUTERR")
...
+else()
+    option(LOG_NO_INFO "Do not Compile Info Log level" ON)
```
Comment (rsanchez15) [inline CMakeLists.txt:286]:
```suggestion
    option(LOG_NO_INFO "Do not compile Info Log level" ON)
```

File: src/cpp/CMakeLists.txt [hunk 10]
```diff
@@ -365,8 +365,6 @@ elseif(NOT EPROSIMA_INSTALLER)
     set_target_properties(${PROJECT_NAME} PROPERTIES VERSION ${PROJECT_VERSION})
     set_target_properties(${PROJECT_NAME} PROPERTIES SOVERSION ${PROJECT_VERSION_MAJOR})
 
-    option(INTERNAL_DEBUG "Activate developer debug messages" OFF)
```
Comment (rsanchez15) [inline src/cpp/CMakeLists.txt:368]: Undo the change

**Review by rsanchez15 — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/participant/RTPSParticipantImpl.cpp [hunk 11]
```diff
@@ -1553,6 +1556,8 @@ bool RTPSParticipantImpl::get_persistence_service(
         {
             logError(RTPS_PARTICIPANT,
                     "Couldn't create writer persistence service for transient/persistent " << debug_label);
+            // Error log level can be disable. Avoid unused warning
+            (void) debug_label;
```
Comment (EduPonz) [inline src/cpp/rtps/participant/RTPSParticipantImpl.cpp:1560]: Remove `debug_label` parameter and correct log accordingly. Currently, the parameter is used when calling the function to log either reader or writer. This information is included in the `EndpointAttributes`

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 12]
```diff
@@ -2242,6 +2242,12 @@ void StatefulWriter::print_inconsistent_acknack(
             << writer_guid << " next SequenceNumber " << next_sequence_number << ". Remote Reader "
             << reader_guid << " requested range is  [" << min_requested_sequence_number
             << ", " << max_requested_sequence_number << "].");
+    // This is necessary to avoid Warning of unused variable in case warning log level is disable
+    (void) writer_guid;
+    (void) reader_guid;
+    (void) min_requested_sequence_number;
+    (void) max_requested_sequence_number;
+    (void) next_sequence_number;
```
Comment (EduPonz) [inline src/cpp/rtps/writer/StatefulWriter.cpp:2250]: `static_cast<void>` is preferred these days. More c++ like

File: src/cpp/rtps/participant/RTPSParticipantImpl.cpp [hunk 13]
```diff
@@ -526,6 +526,9 @@ bool RTPSParticipantImpl::preprocess_endpoint_attributes(
         }
     }
 
+    // Error log level can be disable. Avoid unused warning
+    (void) debug_label;
```
Comment (EduPonz) [inline src/cpp/rtps/participant/RTPSParticipantImpl.cpp:530]: This can go as well

**Review by EduPonz — CHANGES_REQUESTED** [review]
Partial review only relating to `debug_label` paramenters

**Review by rsanchez15 — DISMISSED** [review]
LGTM

File: src/cpp/CMakeLists.txt (hunk 10 above)
Comment (jparisu) [reply src/cpp/CMakeLists.txt]: :+1:

File: CMakeLists.txt (hunk 9 above)
Comment (jparisu) [reply CMakeLists.txt]: :+1:

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3542/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/732/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3143/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3574/)

File: src/cpp/rtps/participant/RTPSParticipantImpl.cpp [hunk 14]
```diff
@@ -449,24 +449,50 @@ RTPSParticipantImpl::~RTPSParticipantImpl()
...
 bool RTPSParticipantImpl::preprocess_endpoint_attributes(
-        const char* debug_label,
         const EntityId_t& entity_id,
         EndpointAttributes& att,
         EntityId_t& entId)
 {
     if (!att.unicastLocatorList.isValid())
     {
-        logError(RTPS_PARTICIPANT, "Unicast Locator List for " << debug_label << " contains invalid Locator");
+        if (att.endpointKind == WRITER)
```
Comment (rsanchez15) [inline src/cpp/rtps/participant/RTPSParticipantImpl.cpp:458]: Leave each of the logs as they were and set up the `debug_label` variable at the beginning of the function. This way you avoid duplicating code. I suggest you use the ternary operator for better code readability.

File: src/cpp/rtps/writer/StatefulWriter.cpp [hunk 15]
```diff
@@ -2243,11 +2243,11 @@ void StatefulWriter::print_inconsistent_acknack(
             << reader_guid << " requested range is  [" << min_requested_sequence_number
             << ", " << max_requested_sequence_number << "].");
     // This is necessary to avoid Warning of unused variable in case warning log level is disable
-    (void) writer_guid;
-    (void) reader_guid;
-    (void) min_requested_sequence_number;
-    (void) max_requested_sequence_number;
-    (void) next_sequence_number;
+    static_cast<void> (writer_guid);
```
Comment (rsanchez15) [inline src/cpp/rtps/writer/StatefulWriter.cpp:2246]:
```suggestion
    static_cast<void>(writer_guid);
```

**Review by rsanchez15 — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/participant/RTPSParticipantImpl.cpp [hunk 16]
```diff
@@ -453,46 +453,21 @@ bool RTPSParticipantImpl::preprocess_endpoint_attributes(
         EndpointAttributes& att,
         EntityId_t& entId)
 {
+    std::string debug_label = (att.endpointKind == WRITER ? "writer" : "reader");
```
Comment (MiguelCompany) [inline src/cpp/rtps/participant/RTPSParticipantImpl.cpp:456]: Using `std::string` will imply unnecessary allocations. Use `const char*` instead.

File: src/cpp/rtps/participant/RTPSParticipantImpl.cpp [hunk 17]
```diff
@@ -1582,42 +1542,29 @@ bool RTPSParticipantImpl::get_persistence_service(
 {
     service = nullptr;
 
+    std::string debug_label = (param.endpointKind == WRITER ? "writer" : "reader");
```
Comment (MiguelCompany) [inline src/cpp/rtps/participant/RTPSParticipantImpl.cpp:1545]: Same here

File: src/cpp/rtps/participant/RTPSParticipantImpl.cpp (hunk 14 above)
Comment (jparisu) [reply src/cpp/rtps/participant/RTPSParticipantImpl.cpp]: :+1:

File: src/cpp/rtps/writer/StatefulWriter.cpp (hunk 15 above)
Comment (jparisu) [reply src/cpp/rtps/writer/StatefulWriter.cpp]: :+1:

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/rtps/participant/RTPSParticipantImpl.cpp (hunk 16 above)
Comment (jparisu) [reply src/cpp/rtps/participant/RTPSParticipantImpl.cpp]: :+1:

File: src/cpp/rtps/participant/RTPSParticipantImpl.cpp (hunk 17 above)
Comment (jparisu) [reply src/cpp/rtps/participant/RTPSParticipantImpl.cpp]: :+1:

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3546/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/736/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3146/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3578/)

**Review by EduPonz — APPROVED** [review]
LGTM
